packages/react-native-reanimated/src/shareables.ts

```diff
@@ -124,176 +192,243 @@ export function makeShareableCloneRecursive<T>(
+  clonedProps.__initData = makeShareableCloneRecursive(
+    value.__initData,
+    true,
+    depth + 1
   );
+
+  for (const [key, element] of Object.entries(value)) {
+    if (key === '__initData' && clonedProps.__initData !== undefined) {
+      continue;
+    }
+    clonedProps[key] = makeShareableCloneRecursive(
+      element,
+      shouldPersistRemote,
+      depth + 1
+    );
+  }
+  const clone = WorkletsModule.makeShareableClone(
+    clonedProps,
+    shouldPersistRemote
+  ) as ShareableRef<T>;
+  shareableMappingCache.set(value, clone);
+  shareableMappingCache.set(clone);
+
+  freezeObjectInDev(value);
+  return clone;
+}
+
+function cloneContextObject<T extends object>(value: T): ShareableRef<T> {
+  const workletContextObjectFactory = (value as Record<string, unknown>)
+    .__workletContextObjectFactory as () => T;
+  const handle = makeShareableCloneRecursive({
+    __init: () => {
+      'worklet';
+      return workletContextObjectFactory();
+    },
+  });
+  shareableMappingCache.set(value, handle);
+  return handle as ShareableRef<T>;
+}
+
+function clonePlainJSObject<T extends object>(
+  value: T,
+  shouldPersistRemote: boolean,
+  depth: number
+): ShareableRef<T> {
+  const clonedProps: Record<string, unknown> = {};
+  for (const [key, element] of Object.entries(value)) {
+    if (key === '__initData' && clonedProps.__initData !== undefined) {
+      continue;
+    }
+    clonedProps[key] = makeShareableCloneRecursive(
+      element,
+      shouldPersistRemote,
+      depth + 1
+    );
+  }
+  const clone = WorkletsModule.makeShareableClone(
+    clonedProps,
+    shouldPersistRemote
+  ) as ShareableRef<T>;
+  shareableMappingCache.set(value, clone);
+  shareableMappingCache.set(clone);
+
+  freezeObjectInDev(value);
+  return clone;
+}
+
+function cloneRegExp<T extends RegExp>(value: T): ShareableRef<T> {
+  const pattern = value.source;
+  const flags = value.flags;
+  const handle = makeShareableCloneRecursive({
+    __init: () => {
+      'worklet';
+      return new RegExp(pattern, flags);
+    },
+  }) as unknown as ShareableRef<T>;
+  shareableMappingCache.set(value, handle);
+
+  return handle;
+}
+
+function cloneError<T extends Error>(value: T): ShareableRef<T> {
+  const { name, message, stack } = value;
+  const handle = makeShareableCloneRecursive({
+    __init: () => {
+      'worklet';
+      // eslint-disable-next-line reanimated/use-reanimated-error
+      const error = new Error();
+      error.name = name;
+      error.message = message;
+      error.stack = stack;
+      return error;
+    },
+  });
+  shareableMappingCache.set(value, handle);
+  return handle as unknown as ShareableRef<T>;
+}
+
+function cloneArrayBuffer<T extends ArrayBuffer>(
+  value: T,
+  shouldPersistRemote: boolean
+): ShareableRef<T> {
+  const clone = WorkletsModule.makeShareableClone(value, shouldPersistRemote);
+  shareableMappingCache.set(value, clone);
+  shareableMappingCache.set(clone);
```

Comment:
This part of the code repeats a few times, so it might be a good idea to move it to a separate function. However, I'm concerned it might make the code less readable. What do you think?

Reply:
In the OG code sometimes there's

```ts
shareableMappingCache.set(value, clone);
shareableMappingCache.set(clone);
```

And sometimes only

```ts
shareableMappingCache.set(value, clone);
```

The reason for this is unclear (I guess somebody just forgot), but let's address it in a separate pull request. Leaving it as is now hints us that there's a problem and that's useful in this case IMO.

Reply:
@piaskowyk Runtime tests behave exactly as on the `main` branch 🥳